core/src/main/java/org/apache/iceberg/BaseRewriteManifests.java

```diff
@@ -312,33 +311,62 @@ long getManifestTargetSizeBytes() {
   }
 
   class WriterWrapper {
-    private ManifestWriter writer;
-
-    synchronized void addEntry(ManifestEntry entry) {
-      if (writer == null) {
-        writer = newWriter();
-      } else if (writer.length() >= getManifestTargetSizeBytes()) {
-        close();
-        writer = newWriter();
+    private final Map<Integer, ManifestWriter> manifestWritersBySpecId = Collections.synchronizedMap(new HashMap<>());
+
+    synchronized void addEntry(ManifestEntry entry, int partitionSpecId) {
+      ManifestWriter manifestWriter = manifestWritersBySpecId.get(partitionSpecId);
+      if (manifestWriter == null) {
+        manifestWriter = newWriter(partitionSpecId);
+      } else if (manifestWriter.length() >= getManifestTargetSizeBytes()) {
+        close(partitionSpecId);
+        manifestWriter = newWriter(partitionSpecId);
       }
-      writer.existing(entry);
+      manifestWriter.existing(entry);
     }
 
-    private ManifestWriter newWriter() {
-      return new ManifestWriter(spec, manifestPath(manifestSuffix.getAndIncrement()), snapshotId());
+    private ManifestWriter newWriter(int partitionSpecId) {
+
+      if (manifestWritersBySpecId.containsKey(partitionSpecId)) {
+        // should not reach here, but checking to avoid resource leak.
+        throw new RuntimeException("ManifestWriter already exists for this partition-specId " + partitionSpecId);
```

Comment:
I think we can avoid the need for this by doing the caching in the same method that creates writers. Instead of checking the writer map before calling this method, this method should handle it:

```java
synchronized void addEntry(ManifestEntry entry, int partitionSpecId) {
  getWriter(partitionSpecId).existing(entry);
}

synchronized ManifestWriter getWriter(int partitionSpecId) {
  ManifestWriter writer = manifestWritersBySpecId.get(partitionSpecId);
  if (writer != null) {
    if (writer.length() < targetSizeBytes) {
      return writer;
    } else {
      close(writer);
    }
  }

  ...
  writer = new ManifestWriter(partitionSpec, outputFile, snapshotId());
  manifestWritersbySpecId.put(partitionSpecId, manifestWriter);
  return manifestWriter;
}
```

That removes the dependency between the methods, which is cleaner than an assertion.